src/parser/lexer.rs

```diff
@@ -1,40 +1,23 @@
-use parser::combinators::{TypeWithLifetime, Str, Unit, Function, Parser, StrParser, Consumer, ParserConsumer, string, character};
-use self::TokenAt::{LParen, RParen, Whitespace, Identifier};
+use parser::combinators::{Parser, StrParser, Consumer, ParserConsumer, string, character};
+use self::Token::{LParen, RParen, Whitespace, Identifier};
 
 #[derive(Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Debug)]
-pub enum TokenAt<'a> {
+pub enum Token<'a> {
     LParen,
     RParen,
     Whitespace,
     Identifier(&'a str),
 }
 
-pub struct Token;
-
-impl<'a> TypeWithLifetime<'a> for Token {
-    type Type = TokenAt<'a>;
-}
-
-impl Function<Unit,Token> for TokenAt<'static> {
-    fn apply<'a>(&self, _: ()) -> TokenAt<'a> {
-        self.clone()
-    }
-}
-
-struct MkIdentifier;
-
-impl Function<Str,Token> for MkIdentifier {
-    fn apply<'a>(&self, name: &'a str) -> TokenAt<'a> {
-        Identifier(name)
-    }
-}
+fn mk_identifier<'a>(s: &'a str) -> Token<'a> { Identifier(s) }
 
 #[allow(non_snake_case)]
-pub fn lexer<C>(consumer: &mut C) where C: ParserConsumer<Str,Token> {
-    let LPAREN = string("(").map(LParen);
-    let RPAREN = string(")").map(RParen);
-    let WHITESPACE = character(char::is_whitespace).map(Whitespace);
-    let IDENTIFIER = character(char::is_alphabetic).and_then(character(char::is_alphanumeric).star()).buffer().map(MkIdentifier);
+pub fn lexer<C,D>(consumer: &mut C) where C: for<'a> ParserConsumer<&'a str,D>, D: for<'a> Consumer<Token<'a>> {
+    let LPAREN = string("(").map(|_| LParen);
+    let RPAREN = string(")").map(|_| RParen);
+    let WHITESPACE = character(char::is_whitespace).map(|_| Whitespace);
+    let IDENTIFIER = character(char::is_alphabetic).and_then(character(char::is_alphanumeric).star())
+                                                   .buffer().map(mk_identifier);
```

Comment:
If I use `Identifier` here, I get an ICE. I will try to reduce and submit a Rust issue.